### PR TITLE
bugfix: Surface errors when a json query arrives at a qe configured to use the graphQL protocol

### PR DESCRIPTION
**Pull request by miguelff**
This PR handles parsing errors gracefully when a JSON query is received by a query engine that is configured with the GraphQL protocol.

### CODE REVIEW

File: query-engine/query-engine/src/server/mod.rs [hunk 1]
```diff
@@ -197,10 +197,20 @@ async fn request_handler(cx: Arc<PrismaContext>, req: Request<Body>) -> Result<R
 
                 Ok(res)
             }
-            Err(_e) => {
+            Err(e) => {
+                let ufe: user_facing_errors::Error = request_handlers::HandlerError::query_conversion(format!(
```
Comment (janpio) [inline query-engine/query-engine/src/server/mod.rs:201]: Is this the only error that can happen in the code above?

File: query-engine/black-box-tests/tests/protocols/mismatched.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
+use crate::helpers::*;
+use query_engine_tests::*;
+
+const JSON_QUERY: &'static str = r###"
```
Comment (janpio) [inline query-engine/black-box-tests/tests/protocols/mismatched.rs:4]: Worth adding the opposite for sending GraphQL to an Engine that expect Json?
(That is the case that happens in PDP Control plane right now when the user has the `jsonProtocol` preview feature activated in their schema)

File: query-engine/query-engine/src/server/mod.rs (hunk 1 above)
Comment (miguelff) [reply query-engine/query-engine/src/server/mod.rs:201]: Yes. That's the catchall. The rest of the errors will be rendered as part of the usual request / response protocol, (i.e. a 200OK response with an `errors` key)

File: query-engine/black-box-tests/tests/protocols/mismatched.rs (hunk 2 above)
Comment (miguelff) [reply query-engine/black-box-tests/tests/protocols/mismatched.rs:4]: For the sake of completion, I can add it. The code catches an error in parsing, no matter the protocol. So it won't cover a different case. Both the graphql and json queries are wrapped in a json object that would be parsed with serde, so the error might be the same, A test to check that will definitely not harm.

**Comment by miguelff** [comment]
After reverting changes to use upstream mobc in https://github.com/prisma/prisma-engines/pull/3875, I had to port the black-box test harness to this PR. Also added a test for graphql queries against an engine configured with the json protocol [here](https://github.com/prisma/prisma-engines/pull/3868/commits/8eed55cd8ec572af52f5fa55aa29fefa4491cff3#diff-6654ee42508933b2956683fb8f5d97fc89e04e9a3709076105ee552fb37f6737R56-R74)

File: query-engine/black-box-tests/tests/protocols/mismatched.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            let client = reqwest::Client::new();
+            let res = client
+                .post("http://0.0.0.0:57582/")
+                .body(JSON_QUERY)
+                .send()
+                .await
+                .unwrap();
+
+            assert_eq!(res.status(), reqwest::StatusCode::UNPROCESSABLE_ENTITY);
+            insta::assert_snapshot!(res.text().await.unwrap(), @r###"{"is_panic":false,"message":"Error parsing Graphql query. data did not match any variant of untagged enum GraphqlBody","backtrace":null}"###);
```
Comment (janpio) [inline query-engine/black-box-tests/tests/protocols/mismatched.rs:91]: Nice how the error message include the kind of query it is expecting and parsing for 👍

**Review by janpio — APPROVED** [review]

**Pull request merged**